packages/editor/src/browser/editor-command.ts

```diff
@@ -333,6 +333,6 @@ export class EditorCommandContribution implements CommandContribution {
         return autoSave === 'on' || autoSave === undefined;
     }
     private async toggleAutoSave(): Promise<void> {
-        this.preferencesService.set(EditorCommandContribution.AUTOSAVE_PREFERENCE, this.isAutoSaveOn() ? 'off' : 'on');
+        this.preferencesService.set(EditorCommandContribution.AUTOSAVE_PREFERENCE, this.isAutoSaveOn() ? 'off' : 'on', PreferenceScope.User);
```

Comment:
From what I understood, the issue with setting the preference with `PreferenceScope.User` means that it will not work for the following use-case which is supported in vscode:

1. `editor.autoSave: false` in the workspace settings (`{workspace}/.theia/settings.json`).
2. attempt to toggle the `autoSave` with the command.
3. the preference value at the workspace scope is not updated while in vscode it is.

Reply:
Hm... so that clarifies where the problem with the current implementation really is. `editor.autoSave` should not be available in folder scope - it should only be set in a local `settings.json` in a single-root workspace, and in a multi-root workspace, only the workspace value should be picked up. If we're picking up a folder value in a multi-root workspace, that's a bug, and I'll look into that. Otherwise, it looks like setting it at workspace level is VSCode's behavior, and would avoid any awkward override situations.

Reply:
@colin-grant-work I might have confused the `workspace` vs `folder` scopes, what I mean is the value under `{workspace}/.vscode/settings.json` is updated when toggling the command which does not happen in Theia if we are to specifically set the scope to `User`.

Reply:
Yes, if there is a bug that is making us pick up folder-scoped preferences, that's the fix to do. Then we should *not* set it in user scope.

Reply:
> From what I understood, the issue with setting the preference with `PreferenceScope.User` means that it will not work for the following use-case which is supported in vscode:
> 
> 1. `editor.autoSave: false` in the workspace settings (`{workspace}/.theia/settings.json`).
> 2. attempt to toggle the `autoSave` with the command.
> 3. the preference value  (`{workspace}/.theia/settings.json`) is not updated while in vscode it is (`.vscode`).

First, we can check if the workspace config contains `editor.autoSave`. If yes, we change the value there. If no, we change it in the user config. Can this work?
 - Users get the Code behavior via the menu or command palette.
 - If users want to modify it manually via the `settings.json`, they know what they're doing.
 - If you're an extension developer, you either want to trigger the default functionality via the command, so we are good. Or you are doing custom things via the preferences service API.

Thoughts? Am I overlooking something?

Reply:
So the code that actually performs this update in VSCode is not the `update` method to which you posted the docs earlier or this thread. It's the `updateValue` method, for which I don't think we have an equivalent. That code looks pretty similar to what you're describing: it loops through all scopes, finding the lowest one in which the value of the given preference isn't currently `undefined` (with user scope as default), and sets the value there. If the value is being deleted, it deletes it in all scopes in which it's defined.

https://github.com/microsoft/vscode/blob/257fbae3f34804eeb427c10bceb4c0ed95870552/src/vs/workbench/services/filesConfiguration/common/filesConfigurationService.ts#L185-L196


https://github.com/microsoft/vscode/blob/257fbae3f34804eeb427c10bceb4c0ed95870552/src/vs/workbench/services/filesConfiguration/common/filesConfigurationService.ts#L185-L196


https://github.com/microsoft/vscode/blob/257fbae3f34804eeb427c10bceb4c0ed95870552/src/vs/workbench/services/configuration/browser/configurationService.ts#L783-L808

If we want to have that logic available for this command, we can implement methods to mirror that in the `PreferenceService`, or we can just do it manually for this command.

Reply:
> we can implement methods to mirror that in the `PreferenceService`

Good idea. Let's keep the new API aligned to the `updateValue` from Code. We can use this new API for auto-save and maybe for others too (theming?).

As an extension developer, I would prefer to use an API that has the same semantics as a VS Code extension API. What do you think?

Reply:
I've implemented this. I'll add some tests next week to make sure it's doing what it should.

Reply:
> I've implemented this. I'll add some tests next week to make sure it's doing what it should.

@vince-fugnitto, the tests are still planned 😃

Reply:
@colin-grant-work sorry about that :)